resolver/parallel_best_resolver.go

```diff
@@ -152,89 +134,149 @@ func (r *ParallelBestResolver) Name() string {
 	return r.String()
 }
 
+// TODO: add resolverCount & retryWithDifferentResolver to output
 func (r *ParallelBestResolver) String() string {
-	result := make([]string, 0, len(r.resolversPerClient))
-
-	for name, res := range r.resolversPerClient {
-		tmp := make([]string, len(res))
-		for i, s := range res {
-			tmp[i] = fmt.Sprintf("%s", s.resolver)
-		}
-
-		result = append(result, fmt.Sprintf("%s (%s)", name, strings.Join(tmp, ",")))
+	result := make([]string, len(r.resolvers))
+	for i, s := range r.resolvers {
+		result[i] = fmt.Sprintf("%s", s.resolver)
 	}
 
-	return fmt.Sprintf("parallel upstreams '%s'", strings.Join(result, "; "))
+	return fmt.Sprintf("%s upstreams '%s (%s)'", parallelResolverType, r.groupName, strings.Join(result, ","))
 }
 
 // Resolve sends the query request to multiple upstream resolvers and returns the fastest result
 func (r *ParallelBestResolver) Resolve(request *model.Request) (*model.Response, error) {
 	logger := log.WithPrefix(request.Log, parallelResolverType)
 
-	var resolvers []*upstreamResolverStatus
-	for _, r := range r.resolversPerClient {
-		resolvers = r
+	if len(r.resolvers) == 1 {
+		logger.WithField("resolver", r.resolvers[0].resolver).Debug("delegating to resolver")
 
-		break
+		return r.resolvers[0].resolver.Resolve(request)
 	}
 
-	if len(resolvers) == 1 {
-		logger.WithField("resolver", resolvers[0].resolver).Debug("delegating to resolver")
+	ctx := context.Background()
```

Comment:
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel() // abort requests to resolvers that lost the race
```

Reply:
What do you mean with previous?
I'm not that familiar with contexts yet, but as far as I know, `cancel` won't result in aborting requests to the lost resolvers, as this context is only used for the `select` `case` and is not passed into any functions.

Reply:
I meant the code currently in main/removed by this PR.
You're right, lets keep the code you wrote for now, and when we use contexts properly add something like this.